File: js_validation/pardox/src/_io/cluster.js
'use strict';

const { getLib } = require('../ffi');
const { DataFrame } = require('../DataFrame');

/**
 * Connect to a cluster of PardoX worker nodes.
 * @param {string[]} addresses  Worker addresses.
 * @returns {number}
 */
function clusterConnect(addresses) {
    const lib = getLib();
    return lib.pardox_cluster_connect(JSON.stringify(addresses));
}

/** Ping a cluster worker. */
function clusterPing(workerName) {
    const lib = getLib();
    return lib.pardox_cluster_ping(workerName);
}

/** Ping all cluster workers. */
function clusterPingAll() {
    const lib = getLib();
    return lib.pardox_cluster_ping_all();
}

/** Checkpoint cluster state to disk. */
function clusterCheckpoint(path) {
    const lib = getLib();
    return lib.pardox_cluster_checkpoint(path);
}

/** Free the cluster connection. */
function clusterFree() {
    const lib = getLib();
    lib.pardox_cluster_free();
}

/** Start a cluster worker node. */
function clusterWorkerStart(port) {
    const lib = getLib();
    return lib.pardox_cluster_worker_start(port);
}

/** Stop the cluster worker. */
function clusterWorkerStop() {
    const lib = getLib();
    return lib.pardox_cluster_worker_stop();
}

/** Register worker with the cluster. */
function clusterWorkerRegister(name, listenAddr) {
    const lib = getLib();
    return lib.pardox_cluster_worker_register(name, listenAddr);
}

/**
 * Scatter a DataFrame across cluster partitions.
 * @param {DataFrame} df          DataFrame to scatter.
 * @param {number}    partitions  Number of partitions.
 * @returns {Object|null}
 */
function clusterScatter(df, partitions) {
    const lib    = getLib();
    const result = lib.pardox_cluster_scatter(df._ptr, partitions);
    return result ? JSON.parse(result) : null;
}

/**
 * Scatter with replication for fault tolerance.
 * @param {DataFrame} df                DataFrame.
 * @param {number}    partitions        Number of partitions.
 * @param {number}    replicationFactor Replication factor.
 * @returns {Object|null}
 */
function clusterScatterResilient(df, partitions, replicationFactor = 2) {
    const lib    = getLib();
    const result = lib.pardox_cluster_scatter_resilient(df._ptr, partitions, replicationFactor);
    return result ? JSON.parse(result) : null;
}

/**
 * Execute SQL query across the cluster.
 * @param {string} sql
 * @returns {DataFrame}
 */
function clusterSql(sql) {
    const lib = getLib();
    const ptr = lib.pardox_cluster_sql(sql);
    if (!ptr) throw new Error('clusterSql returned null.');
    return new DataFrame(ptr);
}

/**
 * Execute fault-tolerant SQL across the cluster.
 * @param {string} sql
 * @returns {DataFrame}
 */
function clusterSqlResilient(sql) {
    const lib = getLib();
    const ptr = lib.pardox_cluster_sql_resilient(sql);
    if (!ptr) throw new Error('clusterSqlResilient returned null.');
    return new DataFrame(ptr);
}

module.exports = {
    clusterConnect,
    clusterPing,
    clusterPingAll,
    clusterCheckpoint,
    clusterFree,
    clusterWorkerStart,
    clusterWorkerStop,
    clusterWorkerRegister,
    clusterScatter,
    clusterScatterResilient,
    clusterSql,
    clusterSqlResilient,
};

File: py_validation/pardox/_io/cluster.py
import json
from ..wrapper import lib


class ClusterConnection:
    """Handle to a cluster of PardoX worker nodes."""

    def __init__(self, ptr):
        self._ptr = ptr

    def ping(self) -> int:
        if not hasattr(lib, 'pardox_cluster_ping'):
            raise NotImplementedError("pardox_cluster_ping not found in Core.")
        return int(lib.pardox_cluster_ping(self._ptr))

    def ping_all(self) -> int:
        if not hasattr(lib, 'pardox_cluster_ping_all'):
            raise NotImplementedError("pardox_cluster_ping_all not found in Core.")
        return int(lib.pardox_cluster_ping_all(self._ptr))

    def sql(self, name: str, query: str):
        from ..frame import DataFrame
        if not hasattr(lib, 'pardox_cluster_sql'):
            raise NotImplementedError("pardox_cluster_sql not found in Core.")
        ptr = lib.pardox_cluster_sql(self._ptr, name.encode('utf-8'), query.encode('utf-8'))
        if not ptr:
            raise RuntimeError(f"cluster.sql({name!r}) returned null.")
        return DataFrame(ptr)

    def checkpoint(self, path: str) -> int:
        if not hasattr(lib, 'pardox_cluster_checkpoint'):
            raise NotImplementedError("pardox_cluster_checkpoint not found in Core.")
        return int(lib.pardox_cluster_checkpoint(self._ptr, path.encode('utf-8')))

    def free(self):
        if self._ptr and hasattr(lib, 'pardox_cluster_free'):
            lib.pardox_cluster_free(self._ptr)
            self._ptr = None

    def __del__(self):
        self.free()

    def __repr__(self):
        return "<PardoX ClusterConnection>"


def cluster_connect(addrs: list):
    """
    Connect to a cluster of PardoX worker nodes.

    Example:
        conn = pdx.cluster_connect(["127.0.0.1:9001"])
        print(conn.ping())
    """
    if not hasattr(lib, 'pardox_cluster_connect'):
        raise NotImplementedError("pardox_cluster_connect not found in Core.")
    ptr = lib.pardox_cluster_connect(json.dumps(addrs).encode('utf-8'))
    if not ptr:
        raise RuntimeError(f"cluster_connect failed for: {addrs}")
    return ClusterConnection(ptr)
